Refactoring in api.rs

khieta opened the pull request:
## Description of changes

This PR performs some refactoring to reduce the size of our api.rs file. For the most part, the changes are just copy-and-paste, but I did need to add some new `pub(crate)` functions to handle the new layer of indirection. I’ll add review comments to any new functions added.
* Moved `EntityUid` and `PolicyId` to a separate api/id.rs file.
* Moved all error definitions to a separate api/err.rs file. (My theory is that this change will make it easier to keep track of the upcoming changes for #745.)
* Moved tests in api.rs to join their friends in the tests.rs file.

The new files are included in api.rs using:
```rust
mod id;
pub use id::*;

mod err;
pub use err::*;
```
I believe this leaves the API unchanged from a user's perspective. Lmk if I'm wrong about that.

The changes in this PR introduce two new clippy warnings:
* [module_name_repetitions](https://rust-lang.github.io/rust-clippy/master/index.html#/module_name_repetitions) for both `PolicyId` and `EntityId` since these types end with "id" (the name of the containing module). I think we can safely ignore these because the types are `pub use`d in api.rs, so users should write `cedar_policy::PolicyId` instead of `cedar_policy::id::PolicyId`.
* [redundant_pub_crate](https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_pub_crate) for the `fold_partition` function. The reason I made it `pub(crate)` instead of private is that it's also used in testing code. I _think_ it is still effectively private with the current setting.

## Issue #, if available

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)

I confirm that this PR (choose one, and delete the other options):

- [X] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

khieta commented on cedar-policy/src/api/id.rs:
```diff
@@ -0,0 +1,289 @@
+        self.0.namespace()
+    }
+
+    /// Construct an [`EntityTypeName`] from the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn new(ty: ast::Name) -> Self {
+        Self(ty)
+    }
```
New function added by this PR

khieta commented on cedar-policy/src/api/id.rs:
```diff
@@ -0,0 +1,289 @@
+        )
+    }
+
+    /// Construct an [`EntityUid`] from the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn new(uid: ast::EntityUID) -> Self {
+        Self(uid)
+    }
+
+    /// Deconstruct an [`EntityUid`] to get the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn into_inner(self) -> ast::EntityUID {
+        self.0
+    }
+
+    /// Deconstruct an [`EntityUid`] to get the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn as_inner(&self) -> &ast::EntityUID {
+        &self.0
+    }
```
New functions added by this PR

khieta commented on cedar-policy/src/api/id.rs:
```diff
@@ -0,0 +1,289 @@
+pub struct PolicyId(ast::PolicyID);
+
+impl PolicyId {
+    /// Construct a [`PolicyId`] from a source string
+    pub fn new(id: impl AsRef<str>) -> Self {
+        Self(ast::PolicyID::from_string(id.as_ref()))
+    }
+
+    /// Deconstruct an [`PolicyId`] to get the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn into_inner(self) -> ast::PolicyID {
+        self.0
+    }
+
+    /// Deconstruct an [`PolicyId`] to get the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn as_inner(&self) -> &ast::PolicyID {
+        &self.0
+    }
```
New functions added by this PR

cdisselkoen replied:
same comment on `as_inner` and `into_inner` here

khieta commented on cedar-policy/src/api.rs:
```diff
@@ -3579,6 +2839,13 @@ impl Expression {
         ))
     }
+
+    /// Deconstruct an [`Expression`] to get the internal type.
+    /// This function is only intended to be used internally.
+    #[cfg(test)]
+    pub(crate) fn into_inner(self) -> ast::Expr {
+        self.0
+    }
```
New function added by this PR

khieta commented on cedar-policy/src/api.rs:
```diff
@@ -3667,6 +2934,13 @@ impl RestrictedExpression {
         ))
     }
+
+    /// Deconstruct an [`RestrictedExpression`] to get the internal type.
+    /// This function is only intended to be used internally.
+    #[cfg(test)]
+    pub(crate) fn into_inner(self) -> ast::RestrictedExpr {
+        self.0
+    }
```
New function added by this PR

khieta commented:
Marking this as ready for review now that (most of) the CI has passed. I don't think the Java failures are the fault of this PR -- I think we just need to review & merge [cedar-java#133](https://github.com/cedar-policy/cedar-java/pull/133). Will re-run CI once that PR is merged.

cdisselkoen commented on cedar-policy/src/api.rs:
```diff
@@ -2124,19 +1557,10 @@ impl EntityUid {
-    }
 }
 
 impl FromStr for EntityUid {
```
Should we move this `FromStr` impl too, since we moved the definition of `EntityUid`?

cdisselkoen commented on cedar-policy/src/api.rs:
```diff
@@ -2688,7 +2040,7 @@ fn is_static_or_link(
 /// Like [`itertools::Itertools::partition_map`], but accepts a function that can fail.
 /// The first invocation of `f` that fails causes the whole computation to fail
-fn fold_partition<T, A, B, E>(
+pub(crate) fn fold_partition<T, A, B, E>(
```
This was recently introduced -- I'm wondering if it's actually used in more than one file?  Or if we can just move it to the place it is used.

khieta replied:
It's currently used (indirectly) by `PolicySet::to_json` and in the tests. One option is to move the relevant tests for this function back to api.rs, how do you feel about that?

cdisselkoen replied:
No strong feelings either way.

cdisselkoen commented on cedar-policy/src/api/err.rs:
```diff
@@ -0,0 +1,569 @@
+    pub err: EvaluationError,
+}
+
+impl From<ast::EntityAttrEvaluationError> for EntityAttrEvaluationError {
```
Let's do a consistency pass for adding `doc(hidden)` to all of these `From` impls that convert from private to public types.  Right now, some do and some do not.  (This can be a followup PR or in this PR)

khieta replied:
Easy enough to do in this PR -- I'll update this thread once I'm done.

khieta replied:
Done 🙂

cdisselkoen commented on cedar-policy/src/api/id.rs:
```diff
@@ -0,0 +1,289 @@
+
+    /// Deconstruct an [`EntityUid`] to get the internal type.
+    /// This function is only intended to be used internally.
+    pub(crate) fn as_inner(&self) -> &ast::EntityUID {
+        &self.0
+    }
```
Thoughts on making this an `impl AsRef` instead of (or in addition to) `as_inner`?

I guess the same goes for `into_inner()` it could be an `impl From` instead / in addition

khieta replied:
Rust noob question: Can I make the `AsRef` and `From` impls `pub(crate)`? I don't want them exposed through `cedar_policy`'s main interface since they expose internal types.

cdisselkoen replied:
I don't believe you can.  We put up with this for `From` impls on error types, but mark them `doc(hidden)`.  We could do the same here.  But I am also perfectly fine with keeping this as `into_inner()` and `as_inner()`, was just asking for thoughts 😄

cdisselkoen reviewed (approved):
Definitely approve of the idea, making api.rs more organized.

khieta commented on cedar-policy/src/api.rs:
```diff
@@ -2124,19 +1557,10 @@ impl EntityUid {
     pub fn from_json(json: serde_json::Value) -> Result<Self, impl miette::Diagnostic> {
```
I left this in api.rs because it can return an error, and I was trying to maintain the hierarchy id.rs -> err.rs -> api.rs. I'm imagining that we'll want to update the return types here to use the types that will be defined (eventually) in err.rs.

Same answer for the comment on the `FromStr` impl below.

cdisselkoen replied:
I'm pretty sure Rust allows cycles in module imports (within the same crate), if that makes a difference :)

john-h-kastner-aws reviewed (approved):
We should add `allow` directives for the new clippy lints if we're OK with them.

khieta commented:
Latest updates:
* Moved `SlotId` to the `id.rs` file (missed it on my first pass)
* Added `#[allow(clippy...)]` to ignore the warnings introduced by my changes
* Moved remaining `EntityUid` and `PolicyId`-related functions to `id.rs`
* Switched to using implementations of `AsRef` and `From` (with `doc(hidden)`) instead of explicit `into_inner` and `as_inner` functions


The only potentially contentious change is this last one. Because `AsRef` was already defined for `PolicyId` using the `str` type, users may need to add annotations to distinguish between the different variants. I needed to do this in one doc comment in my last commit.

andrewmwells-amazon reviewed (approved):
LGTM
